In [67]:
#Anders van Ommeren - 0887404
#Wouter Heemstra
import matplotlib.pyplot as plt
import numpy as np

Geef een uitdrukking voor de relatieve fout $|y - \text{fl}(y)|/|y|$ voor de volgende berekeningen en leg uit waarom we geen nauwkeurig resultaat kunnen verwachten wanneer we de berekening letterlijk uitvoeren in Python.

1. $y = x^2 - z^2$ met $||x| - |z||\ll 1$
1. $y = \sqrt{x + 1} - 1$ met $|x| \ll 1$

Je hoeft geen rekening te houden met representatiefouten van $x$ en $z$ en je mag aannemen dat de relatieve fout in het evalueren van $\sqrt{\cdot}$ is begrensd door de afrondeenheid (dwz dat $\text{fl}(\sqrt{\text{fl}(x)}) = \sqrt{\text{fl}(x)}(1 + \epsilon)$ met $|\epsilon| \leq \eta$).

We gaan er vanuit dat, omdat we geen rekening hoeven te houden met de representatiefouten van $x$ en $z$, dat we $x$ en $z$ als afgeronde reële getallen mogen zien.

We bekijken eerst $y=x^2-z^2$.

Merk op dat $fl(y)=fl(x^2-z^2)=fl(fl(x\cdot x)-fl(z\cdot z))=$

$(x^2(1+\epsilon_1)-z^2(1+\epsilon_2))(1+\epsilon_3)=x^2(1+\epsilon_1)(1+\epsilon_3)-z^2(1+\epsilon_2)(1+\epsilon_3)$, met $\epsilon_1,\epsilon_2,\epsilon_3\in\mathbb{R}$ gekozen zodat $|\epsilon_1|,|\epsilon_2|,|\epsilon_3|\leq \mu$. Volgens Theorem $1.4.1$ vinden we $\xi,\xi'\in\mathbb{R}$ met $|\xi|,|\xi'|\leq\frac{2\mu}{1-2\mu}$, (we schrijven hier $\frac{n\mu}{1-n\mu}=\frac{2\mu}{1-2\mu}$ omdat we met de verzamelingen $\{\epsilon_1,\epsilon_3\}$ en $\{\epsilon_2,\epsilon_3\}$ werken, die beide twee elementen bevatten die voldoen aan de eis in de Theorem, namelijk $|\epsilon_i|\leq\mu$ voor alle $i\in\{1,2,3\}$), en 

$(1+\epsilon_1)(1+\epsilon_3)=(1+\xi)$ en $(1+\epsilon_2)(1+\epsilon_3)=(1+\xi')$.

Dan is $fl(y)=x^2(1+\epsilon_1)(1+\epsilon_3)-z^2(1+\epsilon_2)(1+\epsilon_3)=x^2(1+\xi)-z^2(1+\xi')$,

dus $|fl(y)-y|=|x^2(1+\xi)-z^2(1+\xi')-x^2+z^2|=|x^2\xi-z^2\xi'|$, en dus is

$\frac{|fl(y)-y|}{|y|}=\frac{|x^2\xi-z^2\xi'|}{|x^2-z^2|}$.

Nu bekijken we $y=\sqrt{x+1}-1$. Het is gegeven dat $|x|\ll 1$, dus $x\in(-1,1)$, dus $\sqrt{x}$ bestaat, dus $\sqrt{x+1}$ bestaat.

Merk op dat $fl(y)=fl(\sqrt{x+1}-1)=fl(fl(\sqrt{fl(x+1)})-1)$.

Kies $\epsilon_1',\epsilon_2',\epsilon_3'\in\mathbb{R}$ met $|\epsilon_1'|,|\epsilon_2'|,|\epsilon_3'|\leq\mu$ zodat $fl(fl(\sqrt{fl(x+1)})-1)=$

$(\sqrt{(x+1)(1+\epsilon_1')}(1+\epsilon_2')-1)(1+\epsilon_3')=(\sqrt{x+1}\sqrt{1+\epsilon_1'}(1+\epsilon_2')-1)(1+\epsilon_3')$.

Er geldt dat er een $\epsilon\in\mathbb{R}$ bestaat zodat $|\epsilon|\leq \mu$ en $\sqrt{1+\epsilon_1'}=(1+\epsilon)$. [?] Net als eerder vinden we door Theorem $1.4.1$ een $\theta\le\frac{3\mu}{1-3\mu}$ zodat $(1+\epsilon)(1+\epsilon_2')(1+\epsilon_3')=(1+\theta)$. Dus $(\sqrt{x+1}\sqrt{1+\epsilon_1'}(1+\epsilon_2')-1)(1+\epsilon_3')=(\sqrt{x+1}(1+\epsilon)(1+\epsilon_2')-1)(1+\epsilon_3')=\sqrt{x+1}(1+\epsilon)(1+\epsilon_2')(1+\epsilon_3')-1-\epsilon_3'=\sqrt{x+1}(1+\theta)-1-\epsilon_3'$

$\sqrt{x+1}+\theta\sqrt{x+1}-1-\epsilon_3'$, dus $fl(y)=\sqrt{x+1}+\theta\sqrt{x+1}-1-\epsilon_3'$, dus $|fl(y)-y|=$

$|\sqrt{x+1}+\theta\sqrt{x+1}-1-\epsilon_3'-\sqrt{x+1}+1|=|\theta\sqrt{x+1}-\epsilon_3'|$, dus $\frac{|fl(y)-y|}{|y|}=\frac{|\theta\sqrt{x+1}-\epsilon_3'|}{|\sqrt{x+1}-1|}$.

We kunnen geen nauwkeurig resultaat verwachten wanneer we de berekeningen van $y$ letterlijk uitvoeren in Python omdat als we berekeningen uitvoeren in Python, reële getallen worden omgezet naar getallen gerepresenteerd door een eindig aantal getallen. Stel bijvoorbeeld dat we werken met base $10$ systeem en een irrationeel getal $d$, en stel dat $d$ een getal is van de vorm $0,0...0d_1d_2d_3...$, met $d_i\in\{0,1,...,9\}$ voor alle $i\geq1$. Dan kan het zomaar zijn dat $fl(d)=0,0...0=0$, en dus als Python gebruiken om met $d$ te rekenen, dan geldt dat $x\cdot d=0$ voor alle $x\in fl(\mathbb{R}):=\{fl(r):r\in\mathbb{R}\}$, terwijl het niet klopt dat $x\cdot d=0$ voor alle $x\in\mathbb{R}$. Dus in dit geval kan Python verkeerde resultaten geven.

Geef voor bovenstaande berekeningen een alternatieve uitdrukking, $\widetilde{y}$, die in exacte aritmetiek equivalent is (dwz $y = \widetilde{y}$) maar in eindige precisie een veel nauwkeuriger resultaat geeft. Onderbouw je keuze door ook hier een uitdrukking te geven voor de relatieve fout.

Kies voor $y=x^2-z^2$, $\widetilde{y}=(x-z)(x+z)$. Dan is $fl(\widetilde{y})=fl((x-z)(x+z))=fl(fl(x-z)fl(x+z))=((x-z)(1+\widetilde{\epsilon_1})(x+z)(1+\widetilde{\epsilon_2}))(1+\widetilde{\epsilon_3})=(x-z)(x+z)(1+\widetilde{\epsilon_1})(1+\widetilde{\epsilon_2})(1+\widetilde{\epsilon_3})=(x-z)(x+z)(1+\widetilde{\xi})=$

$(x-z)(x+z)+\widetilde{\xi}(x-z)(x+z)$, met $\widetilde{\epsilon_i}\in\mathbb{R}$ gekozen zodat $|\widetilde{\epsilon_i}|\leq\mu$ voor alle $i\in\{1,2,3\}$ en $\widetilde{\xi}$ het gekozen getal uit Theorem $1.4.1$ zodat $|\widetilde{\xi}|\leq\frac{3\mu}{1-3\mu}$.

We vinden dat $|fl(\widetilde{y})-\widetilde{y}|=|(x-z)(x+z)+\widetilde{\xi}(x-z)(x+z)-(x-z)(x+z)|=|\widetilde{\xi}(x-z)(x+z)|=|\widetilde{\xi}||(x-z)(x+z)|$. Dus $\frac{|fl(\widetilde{y})-\widetilde{y}|}{|\widetilde{y}|}=\frac{|\widetilde{\xi}||(x-z)(x+z)|}{|(x-z)(x+z)|}=|\widetilde{\xi}|$. In dit geval vinden we dus een constante relatieve fout. Eerder vonden we voor $y=x^2-z^2$ dat $\frac{|fl(y)-y|}{|y|}=\frac{|x^2\xi-z^2\xi'|}{|x^2-z^2|}$. Hier kunnen de waarden voor $x$ en $z$ gevarieerd worden en kan de fout eventueel groter gemaakt worden dan $|\widetilde{\xi}|$, wat laat zien dat door $\widetilde{y}$ een nauwkeuriger resultaat geeft dan $y$.

Kies voor $y=\sqrt{x+1}-1$, $\widetilde{y}=\frac{x}{\sqrt{x+1}+1}$. We vinden weer dat, omdat $|x|\ll1$, dat $\sqrt{x+1}$ bestaat, en dat ${\sqrt{x+1}+1}\ne0$. Merk dus op dat, inderdaad, $\frac{x}{\sqrt{x+1}+1}=\frac{(x+1)-1}{\sqrt{x+1}+1}=\frac{(\sqrt{x+1}-1)(\sqrt{x+1}+1)}{\sqrt{x+1}+1}=\sqrt{x+1}-1$.

Nu is $fl(\widetilde{y})=fl(\frac{x}{\sqrt{x+1}+1})=fl(\frac{fl(x)}{fl(fl(\sqrt{fl(x+1)})+1)})=fl(\frac{x}{fl(fl(\sqrt{fl(x+1)})+1)})$.

Nu kiezen we $\widetilde{\epsilon_1'},\widetilde{\epsilon_2'},\widetilde{\epsilon_3'},\widetilde{\epsilon_4'}\in\mathbb{R}$ met $|\widetilde{\epsilon_i'}|\leq\mu$ voor alle $i\in\{1,2,3,4\}$ zodat $fl(\frac{x}{fl(fl(\sqrt{fl(x+1)})+1)})=(\frac{x}{(\sqrt{(x+1)(1+\widetilde{\epsilon_1'})}(1+\widetilde{\epsilon_2'})+1)(1+\widetilde{\epsilon_3'})})(1+\widetilde{\epsilon_4'})=\frac{x(1+\widetilde{\epsilon_4'})}{(\sqrt{x+1}\sqrt{1+\widetilde{\epsilon_1'}}(1+\widetilde{\epsilon_2'})+1)(1+\widetilde{\epsilon_3'})}$. Kies net zoals eerder een $\widetilde{\epsilon}\in\mathbb{R}$ met $|\widetilde{\epsilon}|\leq\mu$ zodat $\sqrt{1+\widetilde{\epsilon_1'}}=(1+\widetilde{\epsilon})$. Dan is $\frac{x(1+\widetilde{\epsilon_4'})}{(\sqrt{x+1}\sqrt{1+\widetilde{\epsilon_1'}}(1+\widetilde{\epsilon_2'})+1)(1+\widetilde{\epsilon_3'})}=\frac{x(1+\widetilde{\epsilon_4'})}{(\sqrt{x+1}(1+\widetilde{\epsilon})(1+\widetilde{\epsilon_2'})+1)(1+\widetilde{\epsilon_3'})}=\frac{x+x\widetilde{\epsilon_4'}}{\sqrt{x+1}(1+\widetilde{\epsilon})(1+\widetilde{\epsilon_2'})(1+\widetilde{\epsilon_3'})+(1+\widetilde{\epsilon_3'})}$. Kies nu net als eerder door Theorem $1.4.1$ een $\widetilde{\theta}\leq\frac{3\mu}{1-3\mu}$ zodat $(1+\widetilde{\epsilon})(1+\widetilde{\epsilon_2'})(1+\widetilde{\epsilon_3'})=(1+\widetilde{\theta})$. Dan is $\frac{x+x\widetilde{\epsilon_4'}}{\sqrt{x+1}(1+\widetilde{\epsilon})(1+\widetilde{\epsilon_2'})(1+\widetilde{\epsilon_3'})+(1+\widetilde{\epsilon_3'})}=\frac{x+x\widetilde{\epsilon_4'}}{\sqrt{x+1}(1+\widetilde{\theta})+1+\widetilde{\epsilon_3'}}$. We vinden dus $fl(\widetilde{y})=\frac{x+x\widetilde{\epsilon_4'}}{\sqrt{x+1}(1+\widetilde{\theta})+1+\widetilde{\epsilon_3'}}=\frac{x+x\widetilde{\epsilon_4'}}{\sqrt{x+1}+\widetilde{\theta}\sqrt{x+1}+1+\widetilde{\epsilon_3'}}$. Dan is $\frac{|fl(\widetilde{y})-\widetilde{y}|}{|\widetilde{y}|}=|\frac{x(1+\widetilde{\epsilon_4'})(\sqrt{x+1}+1)-x(\sqrt{x+1}+\widetilde{\theta}\sqrt{x+1}+1+\widetilde{\epsilon_3'})}{(\sqrt{x+1}+\theta\sqrt{x+1}+1+\widetilde{\epsilon_3'})(\sqrt{x+1}+1)}\frac{\sqrt{x+1}+1}{x}|=|\frac{(1+\widetilde{\epsilon_4'})(\sqrt{x+1}+1)}{\sqrt{x+1}+\widetilde{\theta}\sqrt{x+1}+1+\widetilde{\epsilon_3'}}-1|=|\frac{\sqrt{x+1}+\widetilde{\epsilon_4'}\sqrt{x+1}+1+\widetilde{\epsilon_4'}-\sqrt{x+1}-\widetilde{\theta}\sqrt{x+1}-1-\widetilde{\epsilon_3'}}{\sqrt{x+1}(1+\widetilde{\theta})+1+\widetilde{\epsilon_3'}}|=|\frac{\sqrt{x+1}(\widetilde{\epsilon_4'}-\widetilde{\theta})+\widetilde{\epsilon_4'}-\widetilde{\epsilon_3'}}{\sqrt{x+1}(1+\widetilde{\theta})+1+\widetilde{\epsilon_3'}}|=\frac{|\sqrt{x+1}(\widetilde{\epsilon_4'}-\widetilde{\theta})+\widetilde{\epsilon_4'}-\widetilde{\epsilon_3'}|}{|\sqrt{x+1}(1+\widetilde{\theta})+1+\widetilde{\epsilon_3'}|}$. Dit geeft een nauwkeuriger resultaat dan $\frac{|fl(y)-y|}{|y|}=\frac{|\theta\sqrt{x+1}-\epsilon_3'|}{|\sqrt{x+1}-1|}$. Het getal in de teller van $\frac{|fl(\widetilde{y})-\widetilde{y}|}{|\widetilde{y}|}$ wordt namelijk door een groter getal gedeeld dan het getal in de teller van $\frac{|fl(y)-y|}{|y|}$.

Geef voor bovenstaande berekeningen een voorbeeld in Python waaruit blijkt dat $\widetilde{y}$ inderdaad nauwkeuriger is wanneer we in eindige precisie rekenen. Maak een grafiek van de relatieve fout over een reeks waarden om het schaalgedrag van de formules te demonstreren in de limiet gevallen $|x|\to|z|$ en $|x|\to 0$. *Hint: je kunt soms testgevallen bedenken waar je het exacte antwoord weet door breuken te gebruiken*.

Voor y=x^2-z^2 kiezen we eerst de waarden x=9.3284398294829547237293 en z=4.2820328409227483.

In [68]:
x=9.3284398294829547237293
z=4.2820328409227483
a=float((x**2)-(z**2))
b=float((x-z)*(x+z))
c=(abs(a-x**2+z**2))/(abs(x**2+z**2))
d=(abs(b-(x-z)*(x+z)))/(abs((x-z)*(x+z)))
print(a)
print(b)
print(c)
print(d)

68.68398440154303
68.68398440154303
3.372116765199431e-17
0.0


We vinden dat in het geval $x=9.3284398294829547237293$ en $z=4.2820328409227483$, de relatieve fout voor $\widetilde{y}$ gelijk is aan $0.0$, terwijl de relatieve fout voor $y$ gelijk is aan $3.372116765199431\cdot10^{-17}>0.0$. Hier geeft $\widetilde{y}$ dus een accurater resultaat dan $y$. Vullen we de waarden $x=16.3284398294829547237293$ en $z=8.2820332848782928409227483$ in in deze code, dan vinden we weer dat de relatieve fout van $\widetilde{y}$ gelijk is aan $0.0$ terwijl de relatieve fout van $y$ gelijk is aan $4.2393883706430633\cdot10^{-17}>0.0$, waaraan we weer zien dat $\widetilde{y}$ een accurater resultaat geeft dan $y$. In de eerste opgave hadden we berekend dat de relatieve fout van $\widetilde{y}$ constant is. We zagen net twee gevallen waarin de relatieve fout van $\widetilde{y}$ gelijk was aan $0.0$. Het moet dus wel gelden dat de relatieve fout van $\widetilde{y}$ constant en gelijk aan $0$ is.

Voor $y=\sqrt{x+1}-1$ kiezen we de waarde $x=9.3284398294829547237293$

In [69]:
import math

x=9.3284398294829547237293
a=float((math.sqrt(x+1))-1)
b=float((x)/((math.sqrt(x+1))+1))
c=(abs(a-((math.sqrt(x+1))-1)))/(abs((math.sqrt(x+1))-1))
d=(abs(b-((x)/((math.sqrt(x+1))+1))))/(abs((x)/((math.sqrt(x+1))+1)))
print(a)
print(b)
print(c)
print(d)

2.2137890144629835
2.213789014462983
0.0
0.0


We zien dat de relatieve fout voor $y$ en voor $\widetilde{y}$ beide gelijk zijn aan $0.0$. Dit lijkt het geval te zijn bij iedere waarde die je voor $x$ in kunt vullen. Hieronder is een lijst gegeven van waarden van $x$ waarvoor de relatieve fouten van $y$ en $\widetilde{y}$ gelijk aan elkaar, en gelijk aan $0.0$ zijn. Dit geeft het idee dat de relatieve fouten van $y$ en van $\widetilde{y}$ altijd constant en gelijk aan $0$ zijn.

$x_1=34982683748273437727899838244.3382948298394872983948$

$x_2=math.pi$

$x_3=math.e$

$x_4=8294982$

$x_5=0.237128$